### DISCUSSION TITLE
Creating Power Fx from Classic Workflows (XAML)

### DISCUSSION THREAD

**Discussion by filcole** [microsoft/Power-Fx#2449] [General]
I am experimenting with the possibility of creating a utility to convert _simple_ classic workflows to Power Fx suitable for use in low code plugins.  Starting from a point of total ignorance I have:

* Attempted parsing the XAML of classic workflows using the WWF libraries (failed)
* Realised that I needed an internal representation of a WF XAML.  Learned that this is known as a AST (Abstract Syntax Tree)
* Found [ANTLR](https://www.antlr.org/) (lexer/parser generator) and attempted to create a grammar for WWF XAML - realised that this was quite a task.
* Consumed some of the [Language Implementation Patterns](https://pragprog.com/titles/tpdsl/language-implementation-patterns/) book by [Terence Parr](https://en.wikipedia.org/wiki/Terence_Parr) (ANTLR author)
* Found that there exists a nice parser and executor of XAML (workflows, BR, rollups) already created by Delegate A/S here (https://github.com/delegateas/XrmMockup/blob/master/src/XrmMockupWorkflow/WorkflowTree.cs) by @magesoe et al, which creates a 'compressed' AST for the XAML.  I had a futile attempt to create Power-Fx directly from this AST.
* Found that there is a lexer/parser that generates an AST for Power Fx within this repo. See 
     https://github.com/microsoft/Power-Fx/blob/caa821f071892d195c2034138d6c00d849a90873/src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs#L599

I think the next steps are to:

1) Determine if there is anything in the Power Fx repo/library that will allow an external application to create an AST for Power Fx and then generate the source code (serialised Power Fx) from the AST. @MikeStall is this currently possible, and exposed by the Power-Fx nuget packages?
2) Find a way to transform the XAML/WF AST to a Power Fx AST.  Need to review chapters 11 to 13 of 'Language Implementation Patterns'.

Is this a sane approach?  Are there any thoughts on the approach, or things that should be done differently?

The following nodes in XAML WF need to be converted:

| XAML Activity | Power Fx equivalent (thoughts) |
| --- | --- |
| CreateVariable         |  Suspect this will be eliminated because this is mostly temporary variables used in XAML |
| SelectFirstNonNull   | maps to `Coalesce()` |
| Arithmetic                | maps to `+ - / *` |
| Aggregate                | Think not needed for classic workflows |
| Concat                      | maps to `Concatenate()` |
| AddTime                  | maps to `DateAdd()` |
| SubtractTime            | maps to `DateAdd()` |
| DiffInTime                | maps to `DateDiff()` |
| Trim                          | maps to `TrimEnds()` |   
| Skip                          | ??? only used in Waiting workflows?  If so, abort conversion, and recommend using cloud flow. |
| GetEntityProperty    | Have to work this one out.  Could be `Lookup()`, could be dot notation |
| ConditionExp           | Maps to an expression within an `If()` statement |
| Condition                 | Maps to `If(conditionexpr, thenstatement, ..., defaultstatement)` |
| LogicalComparison  | Suspect this will be eliminated and mapped into ConditionExp |
| SetEntityProperty     | Sets a value (in memory) for use later in CreateEntity or UpdateEntity |
| SetAttributeValue     | ??? |
| CreateEntity              | Maps to `Patch()` when creating a new entity for any `SetAttributeValue` |
| UpdateEntity             | Maps to `Set()` when updating the current record otherwise maps to `Patch()`.  Uses in memory values set by SetEntityProperty |  
| Assign                        | ??? think used internally to set a value to a variable.  Should disappear - or maybe need to convert using `With()` |
| AssignEntity              | Think maps to `Set(Owner, xxx)` or `Patch(table, createdrecord, { Owner: xxx})` |
| ActivityList                 | ???? only used with SendEmail? |
| TerminateWorkflow   | Likely needs to be remapped to `if(not(condition),  XXX)`.  Also need to roll-back any earlier CreateEntity/UpdateEntity if termination is 'Cancelled' instead of 'Successful'.  Use `Error()` if termination is 'Cancelled' to expose message to user. Probably the most difficult to handle?? |
| RollUp                        | Think not needed for classic workflows |
| CallCodeActivity        | Terminate conversion if encountering a CWA |
| ConvertType              | Unsure - may map to `Value()` or `Text()` |
| SendEmail                  | Unsure, perhaps `Patch(email, { contents})` |

**Comment by MikeStall** [comment]
Re 1 - we don't have any programmatic support for ASTs (although it is definitely something we're interested in). 
Just generate the power fx as a string.

**Reply by filcole** [reply]
Thanks @MikeStall.  Yes, I noticed that all the AST classes have internal constructors.  I'm currently got a few classes which create a cut down AST of the necessary PowerFx functions, `If`, `Coalesce`, `Set`, etc, and outputting a Power Fx string from that.  Long statements aren't currently broken over multiple lines :/   I noticed there's PrettyPrinting functions in the repo, but it's not exposed externally.

**Comment by MikeStall** [comment]
Also - re AddTime, etc.... PowerFx has a DateAdd() method. 
https://learn.microsoft.com/en-us/power-platform/power-fx/reference/function-dateadd-datediff

**Reply by MikeStall** [reply]
Also note that PrettyPrinting  shows up in TexlNode.ToString(), which is public :)

**Comment by rnwood** [comment]
Sharing my attempt at the same thing:
https://github.com/rnwood/workflow-moderniser

Interesting to read the conversation above having figured out some of this out independently.

Parsing the XAML is quite straightforward using the `System.Xaml` and `System.Activities` namespaces along with `Microsoft.Xrm.Sdk.Workflow.Activities` from the SDK.

The hairy bit currently is the VB expressions that are embedded. I could not figure out how to successfully parse those despite the above libraries supporting it on paper. So currently, those bits are just parsing out specifics with regex/split/etc.
